src/dataflow/src/render/upsert.rs

```diff
@@ -398,18 +401,67 @@ where
                                     // We store errors as well as non-None values, so that they can be
                                     // retracted if new rows show up for the same key.
                                     let new_value = decoded_value.transpose();
+
+                                    use dataflow_types::plan::Permutation;
+                                    let new_value = new_value.map(|actual_value| {
+                                        actual_value.map(|non_err_val| {
+                                            let total_len = non_err_val.iter().count();
+                                            let (permutation, thinning) =
+                                                Permutation::construct_from_columns(
+                                                    &(0..decoded_key.iter().count())
+                                                        .into_iter()
+                                                        .collect::<Vec<usize>>(),
+                                                    total_len,
+                                                );
+                                            row_packer.clear();
+
+                                            row_packer.extend(
+                                                non_err_val
+                                                    .iter()
+                                                    .enumerate()
+                                                    .filter(|(i, _)| thinning.contains(i))
```

Comment:
I suspect it is the case, but could use @antiguru to confirm, that what we want here is `thinning.map(|index| datums[index])`. The potentially important difference being that `thinning` gets to define the order of things. Unpacking `non_err_val` once would also then allow time linear in that and `thinning`, whereas the above is proportional to their product.

Reply:
Looks like the current implementation ensures that `thinning` is in order, and does not contain repetitions. So, this is probably not wrong, but is a potential performance gotcha.

Reply:
As the value is a packed row, it may be quadratic to index like this, unless I unpack the row into a temporary vec first, thoughts?

Reply:
You can either unpack into a reused `DatumVec`, or pre-determine the "skips" that `thinning` will induce and use an iterator instead. We have done this in the past, in `reduce.rs`, but I think that code might have been a casualty of the permutation work.